target/arm/translate-cheri.inc.c

```diff
@@ -0,0 +1,2045 @@
+                    break;
+                default:
+                    assert(0);
+                }
+                if (extend_size == 2) {
+                    tcg_gen_andi_i64(tcg_rd, tcg_rd, 0xFFFFFFFF);
+                }
+            }
+        } else {
+            if (is_load) {
+                do_fp_ld(ctx, rd, checked, size);
+            } else {
+                do_fp_st(ctx, rd, checked, size);
+            }
+        }
+    }
+
+    if (acquire_release) {
+        TCGBar bar = TCG_MO_ALL;
+        if ((acquire_release == 2) || is_load)
+            bar |= TCG_BAR_LDAQ;
+        if ((acquire_release == 2) || !is_load)
+            bar |= TCG_BAR_STRL;
+        tcg_gen_mb(bar);
+    }
+
+    // Add immediate afterwards if post incrementing
+    if (post_inc && imm != 0) {
+        tcg_gen_addi_i64(wb, wb, imm);
+    }
+
+    if (post_inc || pre_inc) {
+        assert(!exclusive); // otherwise this temp will have died
+        set_gpr_reg_addr_base(ctx, rn, wb, capability_base);
+        tcg_temp_free_i64(wb);
+    }
+
+    if (rn == REG_NONE) {
+        tcg_temp_free_i64(addr);
+    }
+
+    return true;
+}
+
+// All the instructions
+
+TRANS_F(ADR)
+{
+    if (a->Rd == 31)
+        return true;
+
+    // The lack of a capabilities enabled check here is intentional
+
+    bool c64 = IS_C64(ctx);
+    bool P_in_immediate = (!a->op) || !c64;
+    bool page_align = a->op;
+    bool sign_extend = !(c64 && a->op && !a->P);
+    bool pcc = !c64 || !a->op || a->P;
+    bool adrdpb = cctlr_set(ctx, CCTLR_ADRDPB);
+
+    // Concat bits together for immediate
+    int width = 18 + 2;
+    uint64_t im = P_in_immediate ? a->P << 18 : 0;
+    if (P_in_immediate)
+        width++;
+    im = (((im | a->immhi) << 2) | a->immlo);
+
+    // Shift if page aligning
+    if (page_align) {
+        im <<= 12;
+        width += 12;
+    }
+
+    // Sign extend
+    if (sign_extend) {
+        im = SE_64(im, width);
+    }
+
+    // Do this early as it contains a branch
+    if (!pcc && adrdpb)
+        gen_ensure_cap_decompressed(ctx, 28);
+
+    TCGv_i64 new_addr = tcg_temp_new_i64();
+
+    // First get the address
+    if (pcc) {
+        // Can just work this out as a constant
+        uint64_t result = im + ctx->pc_curr;
+        if (!c64 && cctlr_set(ctx, CCTLR_PCCBO)) {
+            result -= ctx->base.pcc_base;
+        }
+        if (page_align) {
+            result &= ~((1 << 12) - 1);
+        }
+        tcg_gen_movi_i64(new_addr, result);
+    } else {
+        if (adrdpb) {
+            gen_cap_get_cursor(ctx, 28, new_addr);
+        } else {
+            tcg_gen_ld_i64(new_addr, cpu_env,
+                           offsetof(CPUARMState, DDC_current) +
+                               offsetof(cap_register_t, _cr_cursor));
+        }
+        tcg_gen_addi_i64(new_addr, new_addr, im);
+        tcg_gen_andi_i64(new_addr, new_addr, ~((1 << 12) - 1));
+    }
+
+    if (c64) {
+        // Derive a capability
```

Comment:
There is no CheckCapabilitiesEnabled() in the ASL. I assume this is the case because ADR(D)P are not actually in the morello encoding space, even if they claim to be?

Reply:
Hm, that seems like a spec bug, unless it's implicitly done elsewhere. As things stand I believe this is a half-hole in the spec; in practice it doesn't matter because you'd have to ERET with SPSR.C64 set (`BX #4` _is_ disabled if capabilities aren't enabled, so userspace can't get into C64 by itself), but it technically allows you to get into a state where it's possible to write back a capability when you shouldn't be able to.

Reply:
Ok, R_VCNGF is the important rule, it's impossible to end up in C64 mode with capability accesses trapped (and ERET will force PSTATE.C64 to 0 regardless of SPSR_ELx.C64 in that case). Guess I should check you've done that... :)

Reply:
is_access_to_capabilities_enabled_at_el is used for ERET, which is copied verbatim from the ASL and will make sure SPSR_C64 is 0. So there is certainly no way of having C64 on ERET if caps not enabled. The other instructions that could set it all do a CheckCapabilitiesEnabled. Think we are OK.

Reply:
:+1: so close to the prize of finding a subtle hole in the architecture..